common/gossmap.c

```diff
@@ -0,0 +1,831 @@
+	if (map_u8(map, message_flags_off) & 1)
+		hc.htlc_max
+			= u64_to_fp16(map_be64(map, htlc_maximum_off), false);
+	else
+		hc.htlc_max = 0xFFFF;
+	hc.base_fee = map_be32(map, fee_base_off);
+	hc.proportional_fee = map_be32(map, fee_prop_off);
+	hc.delay = map_be16(map, cltv_expiry_delta_off);
+
+	/* Check they fit */
+	if (hc.base_fee != map_be32(map, fee_base_off)
+	    || hc.proportional_fee != map_be32(map, fee_prop_off)
+	    || hc.delay != map_be16(map, cltv_expiry_delta_off)) {
+		warnx("channel_update %s ignored: fee %u/%u cltv %u too large",
+		      type_to_string(tmpctx, struct short_channel_id, &scid),
+		      map_be32(map, fee_base_off),
+		      map_be32(map, fee_prop_off),
+		      map_be16(map, cltv_expiry_delta_off));
+		return;
+	}
+
+	chanflags = map_u8(map, channel_flags_off);
+	hc.enabled = !(chanflags & 2);
+	/* Preserve this */
+	hc.nodeidx = chan->half[chanflags & 1].nodeidx;
+	chan->half[chanflags & 1] = hc;
+}
+
+static void remove_channel_by_deletemsg(struct gossmap *map, size_t del_off)
+{
+	struct short_channel_id scid;
+	struct gossmap_chan *chan;
+
+	/* They can delete things we don't know about, since they also
+	 * get their length marked with the deleted bit */
+	/* Note that first two bytes are message type */
+	scid.u64 = map_be64(map, del_off + 2);
+	chan = gossmap_find_chan(map, &scid);
+	if (!chan)
+		return;
+
+	gossmap_remove_chan(map, chan);
+}
+
+struct short_channel_id gossmap_chan_scid(const struct gossmap *map,
+					  const struct gossmap_chan *c)
+{
+	struct short_channel_id scid;
+	scid.u64 = map_be64(map, c->scid_off);
+
+	return scid;
+}
+
+/* BOLT #7:
+ * 1. type: 257 (`node_announcement`)
+ * 2. data:
+ *    * [`signature`:`signature`]
+ *    * [`u16`:`flen`]
+ *    * [`flen*byte`:`features`]
+ *    * [`u32`:`timestamp`]
+ *    * [`point`:`node_id`]
+ *    * [`3*byte`:`rgb_color`]
+ *    * [`32*byte`:`alias`]
+ *    * [`u16`:`addrlen`]
+ *    * [`addrlen*byte`:`addresses`]
+ */
+static void node_announcement(struct gossmap *map, size_t nann_off)
+{
+	const size_t feature_len_off = 2 + 64;
+	size_t feature_len;
+	struct gossmap_node *n;
+	struct node_id id;
+
+	feature_len = map_be16(map, nann_off + feature_len_off);
+	map_nodeid(map, nann_off + feature_len_off + 2 + feature_len + 4, &id);
+	n = gossmap_find_node(map, &id);
+	n->nann_off = nann_off;
+}
+
+static bool map_catchup(struct gossmap *map)
+{
+	size_t reclen;
+	bool changed = false;
+
+	for (; map->map_end + sizeof(struct gossip_hdr) < map->map_size;
+	     map->map_end += reclen) {
+		struct gossip_hdr ghdr;
+		size_t off;
+		u16 type;
+
+		map_copy(map, map->map_end, &ghdr, sizeof(ghdr));
+		reclen = (be32_to_cpu(ghdr.len)
+			  & ~(GOSSIP_STORE_LEN_DELETED_BIT|
+			      GOSSIP_STORE_LEN_PUSH_BIT))
+			+ sizeof(ghdr);
+
+		if (be32_to_cpu(ghdr.len) & GOSSIP_STORE_LEN_DELETED_BIT)
+			continue;
+
+		/* Partial write, this can happen. */
+		if (map->map_end + reclen > map->map_size)
+			break;
```

Comment:
This should only ever happen in non-mmap mode though, right? With `mmap` I'd expect the write of `map->map_size` to be atomic and happen after the update succeeded.

Reply:
mmap or no, it can happen in practice :(